### PR TITLE
feat(python): add 3.9.0

### PR DESCRIPTION
**Pull request by busunkim96**
Add 3.9 and use the latest patch version for other releases.

* Fetch keys in a separate 'RUN' command so errors are not swallowed.
* Use 18.04 as base Ubuntu Image
* Always use gpg keys for installing Python
* Use pip 20.2.4
* gcloud 318.0.0
* Delete unused `silver-python2` image
* Delete makefiles

### CODE REVIEW

**Comment by tmatsuo** [comment]
As I said before, the current dockerfile doesn't abort upon failing to fetch the GPG keys. As a result, it will upload the broken Docker image to the container resistry.

I think you should do either
1. add a test to make sure the image has intended python version
or
2. make the Dockerfile abort when failing to fetch the GPG keys

Otherwise, it is possible we'll have broken docker image uploaded.

Lastly, I still think we should move away from this central docker image because there's no way to make sure the change won't break other repo's tests.

**Comment by busunkim96** [comment]
> As I said before, the current dockerfile doesn't abort upon failing to fetch the GPG keys. As a result, it will upload the broken Docker image to the container resistry.
> 
> I think you should do either
> 
> 1. add a test to make sure the image has intended python version
>    or
> 2. make the Dockerfile abort when failing to fetch the GPG keys
> 
> Otherwise, it is possible we'll have broken docker image uploaded.
> 
Ack, I'll add something that confirms all the Python versions are in the image.

> Lastly, I still think we should move away from this central docker image because there's no way to make sure the change won't break other repo's tests.

Having the Dockerfile in the repo works well for the python-docs-samples image works because there's a critical mass of samples in that repo. There is no equivalent for libraries. If we add a separate Dockerfile to each repo we would be publishing 50+ images that would largely look the same.

I hear the concern about unintended breaks though. As it stands it's unclear which dependencies are needed by all the libraries and which are only needed by 1 or 2. Do you have other thoughts on how to better organize this?

**Comment by tmatsuo** [comment]
> Ack, I'll add something that confirms all the Python versions are in the image.

If you go that route, you should add the test step in the cloudbuild config file.

**Comment by tswast** [comment]
> Do you have other thoughts on how to better organize this?

(Half-kidding) We could always use conda in cases where C/C++ dependencies are required.

**Comment by tmatsuo** [comment]
> There is no equivalent for libraries. If we add a separate Dockerfile to each repo we would be publishing 50+ images that would largely look the same.

Actually this is intended.

We can still have centralized Docker image uploaded to centralized place.
Then each library repo should have almost empty Dockerfiles.

```Dockerfile
from: centralized-image
# Do nothing for now
```

When each library require some dependency, they can add it to THEIR Dockerfile.

This way, the change won't break other libraries. Also the additional build time will be short.

**Review by tmatsuo — APPROVED** [review]
Thanks, definitely better

**Comment by busunkim96** [comment]
@crwilcox Do you recall why this line was added? Are there redis tests depending on this?


https://github.com/googleapis/testing-infra-docker/blob/414b89ce96edea5ac271f73ac768297c88f7670c/python/googleapis/python-multi/Dockerfile#L127-L128
https://github.com/googleapis/testing-infra-docker/commit/414b89ce96edea5ac271f73ac768297c88f7670c

```
Step #4 - "python-multi": [91msudo[0m[91m: [0m[91msystemctl: command not found[0m[91m
Step #4 - "python-multi": The command '/bin/sh -c sudo systemctl enable redis-server.service' returned a non-zero code: 1
Finished Step #4 - "python-multi"
ERROR
ERROR: build step 4 "gcr.io/cloud-builders/docker" failed: step exited with non-zero status: 1
Step #4 - "python-multi": [0m
```

**Comment by crwilcox** [comment]
@busunkim96 I think ndb?

**Comment by busunkim96** [comment]
I had to make more changes to get the images to work correctly. @tmatsuo could you take another look? Thank you!

**Comment by busunkim96** [comment]
Friendly ping @tmatsuo

**Review by tmatsuo — APPROVED** [review]

**Comment by tmatsuo** [comment]
@busunkim96 Sorry for the delay

File: python/googleapis/python-multi/Dockerfile [hunk 1]
```diff
@@ -66,6 +68,27 @@ RUN apt-get update \
...
+RUN set -ex \
+  && gpg --keyserver ha.pool.sks-keyservers.net --recv-keys \
+      # 2.7.17 (Benjamin Peterson)
+      C01E1CAD5EA2C4F0B8E3571504C367C218ADD4FF \
+      # 3.4.10, 3.5.10 (Larry Hastings)
+      97FC712E4C024BBEA48A61ED3A5CA953F73C700D \
+      # 3.6.12, 3.7.9 (Ned Deily)
+      0D96DF4D4110E5C43FBFB17F2D347EA6AA65421D \
+      # 3.8.6 and 3.9.0 (Łukasz Langa)
+      E3FF2839C048B25C084DEBE9B26995E310250568
```
Comment (crwilcox) [inline python/googleapis/python-multi/Dockerfile:90]: dupe comment for 3.10, 3.11

**Review by crwilcox — APPROVED** [review]

File: python/cloud-devrel-kokoro-resources/python-base/Dockerfile [hunk 2]
```diff
@@ -53,26 +55,67 @@ RUN apt-get update \
...
+# See https://www.python.org/downloads/ 'OpenPGP Public Keys' for new keys.
+RUN set -ex \
+  && gpg --keyserver ha.pool.sks-keyservers.net --recv-keys \
+      # 2.7.17 (Benjamin Peterson)
+      C01E1CAD5EA2C4F0B8E3571504C367C218ADD4FF \
+      # 3.4.10, 3.5.10 (Larry Hastings)
+      97FC712E4C024BBEA48A61ED3A5CA953F73C700D \
+      # 3.6.12, 3.7.9 (Ned Deily)
+      0D96DF4D4110E5C43FBFB17F2D347EA6AA65421D \
+      # 3.8.6 and 3.9.0 (Łukasz Langa)
```
[the remark this answers is not in the document]
Comment (tseaver) [reply python/cloud-devrel-kokoro-resources/python-base/Dockerfile:76]: Per the ["OpenPGP Public Keys" section on the Python downloads page](https://www.python.org/downloads/), the release manager is: [Pablo Galindo Salgado (3.10.x and 3.11.x source files and tags)](https://keybase.io/pablogsal/) (key id: [64E628F8D684696D](https://keybase.io/pablogsal/pgp_keys.asc?fingerprint=a035c8c19219ba821ecea86b64e628f8d684696d))

[the remark this answers is not in the document]
Comment (busunkim96) [reply python/cloud-devrel-kokoro-resources/python-base/Dockerfile:76]: 👍  Added.

**Pull request merged**